Upgrade build processes and dependencies

carlospaelinck opened the pull request:
* Upgrade to Webpack 4 and Babel 7
* Use `builder` and `babel-cli` to make ES and CommonJS outputs
* Remove `lib` from source
* Update development demo build and distribution Webpack build
* Add support for Formidable ESLint React ES6 config
* Refactor string-based `ref`s to function-based ones
* Update the Nuka source to conform to ESLint rules
* Add support for Prettier
* Add `yarn.lock` which makes this PR look much bigger than it is
* Bump version number to 4.0.0

Partially addresses #282 
A refactoring of tests will be in a subsequent PR

cc @ryan-roemer @kenwheeler

ryan-roemer commented on .eslintrc:
```diff
@@ -1,75 +1,27 @@
+  "env": {
+    "node": true,
+    "browser": true,
+    "mocha": true
```
Maybe remove this and add a `test/.eslintrc` that extends this one plus adds it there?

Or, if bloating this PR, just separately ticket for later? (Not a huge priority)

carlospaelinck replied:
I am going to remove all test stuff and start fresh. But keep the spec files to ensure we're maintaining coverage when we move to jest.

ryan-roemer replied:
Sounds good!

ryan-roemer commented on package.json:
```diff
@@ -1,31 +1,40 @@
+    "eslint-plugin-import": "^2.9.0",
+    "eslint-plugin-prettier": "^2.6.0",
     "eslint-plugin-react": "^7.4.0",
     "gulp": "^3.9.0",
```
Can we remove all the gulp deps and the`Gulpfile`? Or is that part of the tests?

carlospaelinck replied:
Yeah, tests don't work at all anymore. We're going to move to jest. So I can just remove that stuff for now.

carlospaelinck commented:
@ryan-roemer I removed the gulp and Karma stuff and added the npmignore.

When I run `npm pack` it does look correct as I see this:

<img width="188" alt="screen shot 2018-03-01 at 4 05 27 pm" src="https://user-images.githubusercontent.com/1738349/36872335-9704cbae-1d6a-11e8-9b62-4796da0d07b0.png">

ryan-roemer commented:
Notes on the CI failures:

- If you're using `webpack4` you need node 6.11.something or above. So drop node4 from the matrix. (Maybe add node9 for future goodness)
- The other node versions are failing for invoking gulp. You probably need to scour all text files for incidences of gulp to remove :)

carlospaelinck commented:
That was my next thing to look at before I merge. Thanks!

carlospaelinck commented:
That was also my reasoning for the version bump because I think the goal is to also move away from mixins and `createReactClass` too in addition to updating tooling.
